src/IO/Observer/Actions/RegisterEvents.hpp

```diff
@@ -0,0 +1,135 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <optional>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "IO/Observer/Actions/ObserverRegistration.hpp"
+#include "IO/Observer/ArrayComponentId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/Tags.hpp"
+#include "IO/Observer/TypeOfObservation.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
+
+namespace observers {
+namespace detail {
+CREATE_HAS_TYPE_ALIAS(observation_registration_tags)
+CREATE_HAS_TYPE_ALIAS_V(observation_registration_tags)
+}  // namespace detail
+
+/*!
+ * \brief Retrieves the observation type and key from an event, if it is an
+ * event that needs to register with the observers.
+ *
+ * The event must define an `observation_registration_tags` type alias that is a
+ * `tmpl::list<>` of all the tags from the DataBox needed to construct the
+ * `ObservationKey`, and a `get_observation_type_and_key_for_registration`
+ * member function if it is to be registered automatically.
+ */
+template <typename EventRegistrars, typename DbTagsList>
+std::optional<
+    std::pair<::observers::TypeOfObservation, ::observers::ObservationKey>>
+get_registration_observation_type_and_key(
+    const Event<EventRegistrars>& event,
+    const db::DataBox<DbTagsList>& box) noexcept {
+  std::optional<
+      std::pair<::observers::TypeOfObservation, ::observers::ObservationKey>>
+      result{};
+  bool already_registered = false;
+  tmpl::for_each<typename Event<EventRegistrars>::creatable_classes>(
+      [&already_registered, &box, &event, &result](auto event_type_v) noexcept {
+        using EventType = typename decltype(event_type_v)::type;
+        if constexpr (detail::has_observation_registration_tags_v<EventType>) {
+          const auto* const derived_class_ptr =
+              dynamic_cast<const EventType*>(&event);
```

Comment:
Can you a a short note of what is going on here with the casting? Is the idea that only one type EventType results in a valid (non-nullptr) result from the cast?